java.lang.NumberFormatException: For input string: "1,146" for custom endpoint and java.net.SocketTimeoutException: Read timed out for spline server

Hi,

spline agent - za.co.absa.spline.agent.spark:spark-2.4-spline-agent-bundle_2.11:0.7.1
spline server - absaoss/spline-rest-server:0.7.1


I am trying to integrate spline agent with spark version 2.4.4 and scala version 2.11.12  and getting below error , and also i see similar issue is fixed as part of #345  and in version 0.7.1

Endpoint URL - **custom** 
```
ERROR SplineQueryExecutionListener: Unexpected error occurred during lineage processing for application: 
java.lang.NumberFormatException: For input string: "1,146"
```

Full Error:
```
ERROR SplineQueryExecutionListener: Unexpected error occurred during lineage processing for application:
java.lang.NumberFormatException: For input string: "1,146"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at java.lang.Integer.parseInt(Integer.java:615)
	at scala.collection.immutable.StringLike$class.toInt(StringLike.scala:273)
	at scala.collection.immutable.StringOps.toInt(StringOps.scala:29)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.modelmapper.ModelMapperV1$.za$co$absa$spline$harvester$dispatcher$httpdispatcher$modelmapper$ModelMapperV1$$toV1OperationId$1(ModelMapperV1.scala:115)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.modelmapper.ModelMapperV1$.za$co$absa$spline$harvester$dispatcher$httpdispatcher$modelmapper$ModelMapperV1$$toV1ReadOperation$1(ModelMapperV1.scala:98)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.modelmapper.ModelMapperV1$$anonfun$toV1Operations$1$1$$anonfun$apply$6.apply(ModelMapperV1.scala:79)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.modelmapper.ModelMapperV1$$anonfun$toV1Operations$1$1$$anonfun$apply$6.apply(ModelMapperV1.scala:79)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.modelmapper.ModelMapperV1$$anonfun$toV1Operations$1$1.apply(ModelMapperV1.scala:79)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.modelmapper.ModelMapperV1$$anonfun$toV1Operations$1$1.apply(ModelMapperV1.scala:79)
	at scala.Option.map(Option.scala:146)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.modelmapper.ModelMapperV1$.toV1Operations$1(ModelMapperV1.scala:79)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.modelmapper.ModelMapperV1$.toDTO(ModelMapperV1.scala:201)
	at za.co.absa.spline.harvester.dispatcher.HttpLineageDispatcher.send(HttpLineageDispatcher.scala:59)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler$$anonfun$onSuccess$2.apply(QueryExecutionEventHandler.scala:45)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler$$anonfun$onSuccess$2.apply(QueryExecutionEventHandler.scala:43)
	at scala.Option.foreach(Option.scala:257)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler.onSuccess(QueryExecutionEventHandler.scala:43)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(SplineQueryExecutionListener.scala:40)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(SplineQueryExecutionListener.scala:40)
	at scala.Option.foreach(Option.scala:257)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener$$anonfun$onSuccess$1.apply$mcV$sp(SplineQueryExecutionListener.scala:40)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.withErrorHandling(SplineQueryExecutionListener.scala:49)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.onSuccess(SplineQueryExecutionListener.scala:39)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(QueryExecutionListener.scala:124)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling$1.apply(QueryExecutionListener.scala:145)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling$1.apply(QueryExecutionListener.scala:143)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
	at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
	at org.apache.spark.sql.util.ExecutionListenerManager.org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling(QueryExecutionListener.scala:143)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply$mcV$sp(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager.readLock(QueryExecutionListener.scala:156)
	at org.apache.spark.sql.util.ExecutionListenerManager.onSuccess(QueryExecutionListener.scala:122)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:678)
	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:285)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:271)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:229)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
```

Appreciate the help to fix this. Thanks!

When tried spline rest server for the same version , getting time out error
```
Cannot send lineage data to http://localhost:8080/producer/execution-plans 

and followed by it

java.net.SocketTimeoutException: Read timed out 
```

Full error -
```
22/05/10 12:37:27 ERROR SplineQueryExecutionListener: Unexpected error occurred during lineage processing for application: 
java.lang.RuntimeException: Cannot send lineage data to http://localhost:8080/producer/execution-plans
	at za.co.absa.spline.harvester.dispatcher.HttpLineageDispatcher.sendJson(HttpLineageDispatcher.scala:85)
	at za.co.absa.spline.harvester.dispatcher.HttpLineageDispatcher.send(HttpLineageDispatcher.scala:60)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler$$anonfun$onSuccess$2.apply(QueryExecutionEventHandler.scala:45)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler$$anonfun$onSuccess$2.apply(QueryExecutionEventHandler.scala:43)
	at scala.Option.foreach(Option.scala:257)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler.onSuccess(QueryExecutionEventHandler.scala:43)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(SplineQueryExecutionListener.scala:40)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(SplineQueryExecutionListener.scala:40)
	at scala.Option.foreach(Option.scala:257)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener$$anonfun$onSuccess$1.apply$mcV$sp(SplineQueryExecutionListener.scala:40)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.withErrorHandling(SplineQueryExecutionListener.scala:49)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.onSuccess(SplineQueryExecutionListener.scala:39)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(QueryExecutionListener.scala:124)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling$1.apply(QueryExecutionListener.scala:145)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling$1.apply(QueryExecutionListener.scala:143)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
	at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
	at org.apache.spark.sql.util.ExecutionListenerManager.org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling(QueryExecutionListener.scala:143)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply$mcV$sp(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager.readLock(QueryExecutionListener.scala:156)
	at org.apache.spark.sql.util.ExecutionListenerManager.onSuccess(QueryExecutionListener.scala:122)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:678)
	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:285)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:271)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:229)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.SocketTimeoutException: Read timed out
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1947)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1942)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1941)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1511)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1495)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	at scalaj.http.HttpRequest.scalaj$http$HttpRequest$$doConnection(Http.scala:367)
	at scalaj.http.HttpRequest.exec(Http.scala:343)
	at scalaj.http.HttpRequest.asString(Http.scala:491)
	at za.co.absa.spline.harvester.dispatcher.httpdispatcher.rest.RestEndpoint.post(RestEndpoint.scala:43)
	at za.co.absa.spline.harvester.dispatcher.HttpLineageDispatcher.sendJson(HttpLineageDispatcher.scala:78)
	... 39 more
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:735)
	at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:678)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1590)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1495)
	at scalaj.http.HttpRequest.scalaj$http$HttpRequest$$doConnection(Http.scala:365)
	... 43 more
```

--- a/core/src/main/scala/za/co/absa/spline/harvester/IdGeneratorsBundle.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/IdGeneratorsBundle.scala
@@ -29,13 +29,14 @@ class IdGeneratorsBundle(execPlanUUIDGeneratorFactory: UUIDGeneratorFactory[UUID
   val operationIdGenerator: IdGenerator[Any, String] = new SequentialIdGenerator(OperationIdTemplate)
   val dataTypeIdGenerator: IdGenerator[Any, UUID] =
     new ComposableIdGenerator(
-      new SequentialIdGenerator("{0}"),
+      new SequentialIdGenerator(NumberTemplate),
       new UUID5IdGenerator[String](UUIDNamespace.DataType)
     )
 }
 
 object IdGeneratorsBundle {
-  val AttributeIdTemplate = "attr-{0}"
-  val ExpressionIdTemplate = "expr-{0}"
-  val OperationIdTemplate = "op-{0}"
+  val NumberTemplate = "{0,number,#}"
+  val AttributeIdTemplate = s"attr-$NumberTemplate"
+  val ExpressionIdTemplate = s"expr-$NumberTemplate"
+  val OperationIdTemplate = s"op-$NumberTemplate"
 }

--- a/core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/modelmapper/ModelMapperV1.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/modelmapper/ModelMapperV1.scala
@@ -111,8 +111,8 @@ object ModelMapperV1 extends ModelMapper {
       )
 
     def toV1OperationId(opIdV11: String): Int = {
-      val Array(opIdV1Str: String) = new MessageFormat(IdGeneratorsBundle.OperationIdTemplate).parse(opIdV11)
-      opIdV1Str.toInt
+      val Array(opIdV1: Number) = new MessageFormat(IdGeneratorsBundle.OperationIdTemplate).parse(opIdV11)
+      opIdV1.intValue()
     }
 
     def toV1SystemInfo(nav: v1_1.NameAndVersion) = v1_0.SystemInfo(nav.name, nav.version)

--- a/core/src/main/scala/za/co/absa/spline/harvester/idGenerators.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/idGenerators.scala
@@ -53,10 +53,11 @@ class ComposableIdGenerator[-A, B, +C](idGen1: IdGenerator[A, B], idGen2: IdGene
 }
 
 abstract class HashBasedUUIDGenerator[-A <: AnyRef](namespace: UUID, hashAlgorithm: String) extends IdGenerator[A, UUID] {
+  private val digest = MessageDigest.getInstance(hashAlgorithm)
+  private val generator = Generators.nameBasedGenerator(namespace, digest)
+
   override def nextId(entity: A): UUID = {
     val input = entity.toJson
-    val digest = MessageDigest.getInstance(hashAlgorithm)
-    val generator = Generators.nameBasedGenerator(namespace, digest)
     generator.generate(input)
   }
 }

--- a/core/src/test/scala/za/co/absa/spline/harvester/IdGeneratorsBundleSpec.scala
+++ b/core/src/test/scala/za/co/absa/spline/harvester/IdGeneratorsBundleSpec.scala
@@ -22,6 +22,8 @@ import org.scalatestplus.mockito.MockitoSugar
 import za.co.absa.spline.harvester.IdGenerator.UUIDGeneratorFactory
 import za.co.absa.spline.producer.model.v1_1.ExecutionPlan
 
+import java.text.MessageFormat
+
 class IdGeneratorsBundleSpec
   extends AnyFlatSpec
     with Matchers
@@ -30,7 +32,6 @@ class IdGeneratorsBundleSpec
   behavior of "dataTypeIdGenerator"
 
   it should "generate deterministic sequence of unique ID" in {
-
     val execPlanUUIDGeneratorFactoryMock = mock[UUIDGeneratorFactory[Any, ExecutionPlan]]
 
     val gen1 = new IdGeneratorsBundle(execPlanUUIDGeneratorFactoryMock).dataTypeIdGenerator
@@ -41,6 +42,26 @@ class IdGeneratorsBundleSpec
 
     seq1 shouldEqual seq2
     seq1.distinct.length shouldBe 10
+  }
+
+  // https://github.com/AbsaOSS/spline-spark-agent/issues/431
+  it should "properly format integers >= 1000 in ID strings" in {
+    val execPlanUUIDGeneratorFactoryMock = mock[UUIDGeneratorFactory[Any, ExecutionPlan]]
+    val idGenBundle = new IdGeneratorsBundle(execPlanUUIDGeneratorFactoryMock)
+
+    val gen1 = idGenBundle.attributeIdGenerator
+    val gen2 = idGenBundle.expressionIdGenerator
+    val gen3 = idGenBundle.operationIdGenerator
+
+    (1 to 1000).foreach(_ => {
+      gen1.nextId()
+      gen2.nextId()
+      gen3.nextId()
+    })
+
+    new MessageFormat(IdGeneratorsBundle.AttributeIdTemplate).parse(gen1.nextId()).head shouldBe 1000
+    new MessageFormat(IdGeneratorsBundle.ExpressionIdTemplate).parse(gen2.nextId()).head shouldBe 1000
+    new MessageFormat(IdGeneratorsBundle.OperationIdTemplate).parse(gen3.nextId()).head shouldBe 1000
   }
 
 }

